src/NHibernate.Test/NHSpecificTest/GH948/EntityProjectionType.cs

```diff
@@ -0,0 +1,92 @@
+﻿using System;
+using System.Data.Common;
+using NHibernate.Engine;
+using NHibernate.Event;
+using NHibernate.Persister.Entity;
+using NHibernate.Type;
+
+namespace NHibernate.Test.NHSpecificTest.GH948
+{
+	/// <summary>
+	/// Specialized type for retrieving entity from Criteria/QueryOver API projection.
+	/// Intended to be used only inside <see cref="EntityProjection"/>
+	/// </summary>
+	[Serializable]
+	internal partial class EntityProjectionType : ManyToOneType
+	{
+		private readonly EntityProjection _projection;
+
+		public EntityProjectionType(EntityProjection projection) : base(projection.RootEntity.FullName, projection.Lazy)
+		{
+			_projection = projection;
+		}
+
+		public override object NullSafeGet(DbDataReader rs, string name, ISessionImplementor session, object owner)
+		{
+			if (_projection.Lazy)
+				return base.NullSafeGet(rs, name, session, owner);
+
+			return GetInitializedEntityFromProjection(rs, session);
+		}
+
+		private object GetInitializedEntityFromProjection(DbDataReader rs, ISessionImplementor session)
+		{
+			var identifier = _projection.Persister.IdentifierType.NullSafeGet(rs, _projection.IdentifierAliases, session, null);
+			if (identifier == null)
+			{
+				return null;
+			}
+
+			var entity = CreateInitializedEntity(
+				rs,
+				session,
+				_projection.Persister,
+				identifier,
+				_projection.PropertyAliases,
+				LockMode.None,
+				_projection.FetchLazyProperties,
+				_projection.IsReadOnly);
+
+			return entity;
+		}
+
+		private static object CreateInitializedEntity(DbDataReader rs, ISessionImplementor session, IQueryable persister, object identifier, string[][] propertyAliases, LockMode lockMode, bool fetchLazyProperties, bool readOnly)
```

Comment:
The main concern is this method. If/when it's properly implemented I think it should be moved to some utility class. As finding a way to properly initialize entity from data reader wasn't fun.